File: go/circuitrelay/relayframe.go
// SPDX-License-Identifier: MIT

// Package circuitrelay implements the native circuit-relay-v2 wire frame — the
// decentralised any-node relay that lets a node reach a peer it cannot contact
// directly by routing through a third node reachable to both. This is the Go
// side of the cross-language protocol; conventions mirror go/dtn (envelope.go):
// version byte first, little-endian integers, uint16-LE length-prefixed UTF-8
// strings, the 16-byte connection id as a UUID in RFC-4122 big-endian order, and
// an int32-LE length-prefixed payload last. Byte-identical to the C# reference
// (AetherNet.CircuitRelay.RelayFrameSerializer) and pinned by fixtures/circuit-relay.
package circuitrelay

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"io"

	"github.com/google/uuid"
)

// FrameVersion is the format-version byte at offset 0 of every relay frame.
const FrameVersion byte = 0x01

const maxPayload = 16 * 1024 * 1024

// MessageType is the circuit-relay-v2 verb.
type MessageType byte

const (
	MsgReserve         MessageType = 1
	MsgReserveResponse MessageType = 2
	MsgConnect         MessageType = 3
	MsgStop            MessageType = 4
	MsgStopResponse    MessageType = 5
	MsgConnectResponse MessageType = 6
	MsgData            MessageType = 7
)

// Status is a relay response result code.
type Status byte

const (
	StatusOk                    Status = 0
	StatusReservationRefused    Status = 1
	StatusNoReservation         Status = 2
	StatusResourceLimitExceeded Status = 3
	StatusPermissionDenied      Status = 4
	StatusConnectionFailed      Status = 5
	StatusMalformedMessage      Status = 6
)

// RelayFrame is a single circuit-relay-v2 wire frame (one fixed layout carries
// every verb, type-discriminated).
type RelayFrame struct {
	Type                   MessageType
	Status                 Status
	SourceUhid             string
	DestinationUhid        string
	RelayUhid              string
	ConnectionID           string // UUID string; "" is treated as the nil UUID
	ReservationExpiresAtMs int64
	LimitDurationSeconds   int32
	LimitDataBytes         int64
	Payload                []byte
}

// Serialize encodes a RelayFrame to its binary wire form.
func Serialize(f *RelayFrame) ([]byte, error) {
	if f == nil {
		return nil, fmt.Errorf("relay: frame must not be nil")
	}
	connID := f.ConnectionID
	if connID == "" {
		connID = uuid.Nil.String()
	}
	idBytes, err := marshalUUID(connID)
	if err != nil {
		return nil, err
	}

	buf := new(bytes.Buffer)
	buf.WriteByte(FrameVersion)
	buf.WriteByte(byte(f.Type))
	buf.WriteByte(byte(f.Status))
	if err := writeStr(buf, f.SourceUhid); err != nil {
		return nil, err
	}
	if err := writeStr(buf, f.DestinationUhid); err != nil {
		return nil, err
	}
	if err := writeStr(buf, f.RelayUhid); err != nil {
		return nil, err
	}
	buf.Write(idBytes)
	writeI64(buf, f.ReservationExpiresAtMs)
	writeI32(buf, f.LimitDurationSeconds)
	writeI64(buf, f.LimitDataBytes)
	if err := writeBytes32(buf, f.Payload); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

// Deserialize decodes a RelayFrame from its binary wire form.
func Deserialize(data []byte) (*RelayFrame, error) {
	r := bytes.NewReader(data)
	if err := expectVersion(r); err != nil {
		return nil, err
	}
	typ, err := r.ReadByte()
	if err != nil {
		return nil, fmt.Errorf("relay: read type: %w", err)
	}
	if typ == 0 || typ > byte(MsgData) {
		return nil, fmt.Errorf("relay: invalid message type %d", typ)
	}
	status, err := r.ReadByte()
	if err != nil {
		return nil, fmt.Errorf("relay: read status: %w", err)
	}
	if status > byte(StatusMalformedMessage) {
		return nil, fmt.Errorf("relay: invalid status %d", status)
	}
	src, err := readStr(r)
	if err != nil {
		return nil, err
	}
	dst, err := readStr(r)
	if err != nil {
		return nil, err
	}
	relay, err := readStr(r)
	if err != nil {
		return nil, err
	}
	connID, err := readUUID(r)
	if err != nil {
		return nil, err
	}
	reservationExpiresAtMs, err := readI64(r)
	if err != nil {
		return nil, err
	}
	limitDurationSeconds, err := readI32(r)
	if err != nil {
		return nil, err
	}
	limitDataBytes, err := readI64(r)
	if err != nil {
		return nil, err
	}
	payload, err := readBytes32(r)
	if err != nil {
		return nil, err
	}
	return &RelayFrame{
		Type:                   MessageType(typ),
		Status:                 Status(status),
		SourceUhid:             src,
		DestinationUhid:        dst,
		RelayUhid:              relay,
		ConnectionID:           connID,
		ReservationExpiresAtMs: reservationExpiresAtMs,
		LimitDurationSeconds:   limitDurationSeconds,
		LimitDataBytes:         limitDataBytes,
		Payload:                payload,
	}, nil
}

// ---- low-level helpers (mirror go/dtn/envelope.go) --------------------------

func marshalUUID(s string) ([]byte, error) {
	id, err := uuid.Parse(s)
	if err != nil {
		return nil, fmt.Errorf("relay: invalid uuid %q: %w", s, err)
	}
	b, err := id.MarshalBinary() // RFC-4122 big-endian, 16 bytes
	if err != nil {
		return nil, fmt.Errorf("relay: marshal uuid: %w", err)
	}
	return b, nil
}

func readUUID(r io.Reader) (string, error) {
	b := make([]byte, 16)
	if _, err := io.ReadFull(r, b); err != nil {
		return "", fmt.Errorf("relay: read uuid: %w", err)
	}
	var id uuid.UUID
	if err := id.UnmarshalBinary(b); err != nil {
		return "", fmt.Errorf("relay: unmarshal uuid: %w", err)
	}
	return id.String(), nil
}

func expectVersion(r *bytes.Reader) error {
	v, err := r.ReadByte()
	if err != nil {
		return fmt.Errorf("relay: read version: %w", err)
	}
	if v != FrameVersion {
		return fmt.Errorf("relay: unsupported frame version 0x%02x", v)
	}
	return nil
}

func writeI32(buf *bytes.Buffer, v int32) { _ = binary.Write(buf, binary.LittleEndian, v) }
func writeI64(buf *bytes.Buffer, v int64) { _ = binary.Write(buf, binary.LittleEndian, v) }

func writeStr(buf *bytes.Buffer, s string) error {
	b := []byte(s)
	if len(b) > 65535 {
		return fmt.Errorf("relay: string too long: %d bytes exceeds 65535", len(b))
	}
	_ = binary.Write(buf, binary.LittleEndian, uint16(len(b)))
	buf.Write(b)
	return nil
}

func writeBytes32(buf *bytes.Buffer, data []byte) error {
	if len(data) > maxPayload {
		return fmt.Errorf("relay: payload too large: %d bytes exceeds %d", len(data), maxPayload)
	}
	_ = binary.Write(buf, binary.LittleEndian, int32(len(data)))
	buf.Write(data)
	return nil
}

func readI32(r io.Reader) (int32, error) {
	var v int32
	err := binary.Read(r, binary.LittleEndian, &v)
	return v, err
}

func readI64(r io.Reader) (int64, error) {
	var v int64
	err := binary.Read(r, binary.LittleEndian, &v)
	return v, err
}

func readStr(r io.Reader) (string, error) {
	var n uint16
	if err := binary.Read(r, binary.LittleEndian, &n); err != nil {
		return "", err
	}
	if n == 0 {
		return "", nil
	}
	b := make([]byte, n)
	if _, err := io.ReadFull(r, b); err != nil {
		return "", err
	}
	return string(b), nil
}

func readBytes32(r io.Reader) ([]byte, error) {
	var n int32
	if err := binary.Read(r, binary.LittleEndian, &n); err != nil {
		return nil, err
	}
	if n < 0 {
		return nil, fmt.Errorf("relay: negative payload length: %d", n)
	}
	if int64(n) > maxPayload {
		return nil, fmt.Errorf("relay: payload length %d exceeds %d", n, maxPayload)
	}
	if n == 0 {
		return []byte{}, nil
	}
	b := make([]byte, n)
	if _, err := io.ReadFull(r, b); err != nil {
		return nil, err
	}
	return b, nil
}

File: typescript/src/circuitrelay/RelayFrame.ts
/**
 * Binary circuit-relay-v2 wire-frame serialization — the cross-language wire
 * format for native (no-libp2p) any-node relaying, carried in MeshPacket.payload
 * the same way the DTN envelope is. Conventions mirror DtnEnvelope /
 * PacketSerializer exactly so the eight language SDKs stay byte-identical (and
 * are pinned by the fixtures/circuit-relay corpus): every frame begins with a
 * single format-version byte (readers reject any other value); all multi-byte
 * integers are little-endian; the 16-byte connectionId is the UUID in RFC-4122
 * big-endian order; strings are uint16-LE length-prefixed UTF-8; the payload is
 * int32-LE length-prefixed raw bytes and is always the last field.
 *
 * Layout (fixed, every field always present):
 *   version u8 | type u8 | status u8
 *   srcUhid u16+utf8 | dstUhid u16+utf8 | relayUhid u16+utf8
 *   connId 16B(BE) | reservationExpiresAtMs i64 | limitDurationSeconds i32 | limitDataBytes i64
 *   payload i32+bytes
 * Minimum size (all strings empty, no payload): 49 bytes.
 *
 * Byte-identical to the C# reference (AetherNet.CircuitRelay.RelayFrameSerializer)
 * and the Go oracle (go/circuitrelay).
 *
 * SPDX-License-Identifier: MIT
 */

/** Circuit-relay-v2 verb (message type). */
export enum RelayMessageType {
  Reserve = 1,
  ReserveResponse = 2,
  Connect = 3,
  Stop = 4,
  StopResponse = 5,
  ConnectResponse = 6,
  Data = 7,
}

/** Relay response result code. */
export enum RelayStatus {
  Ok = 0,
  ReservationRefused = 1,
  NoReservation = 2,
  ResourceLimitExceeded = 3,
  PermissionDenied = 4,
  ConnectionFailed = 5,
  MalformedMessage = 6,
}

/**
 * A single circuit-relay-v2 wire frame. One fixed layout carries every verb,
 * type-discriminated. `reservationExpiresAtMs` and `limitDataBytes` are i64 on
 * the wire but carried as JS numbers (matching how DtnEnvelope handles
 * createdAtMs); values stay well within Number.MAX_SAFE_INTEGER.
 */
export interface RelayFrame {
  type: RelayMessageType;
  status: RelayStatus;
  sourceUhid: string;
  destinationUhid: string;
  relayUhid: string;
  /** UUID string; "" (or omitted) is treated as the nil UUID (16 zero bytes). */
  connectionId: string;
  reservationExpiresAtMs: number;
  limitDurationSeconds: number;
  limitDataBytes: number;
  payload: Uint8Array;
}

export const RELAY_FRAME_VERSION = 0x01;
const MAX_RELAY_PAYLOAD = 16 * 1024 * 1024;
const NIL_UUID = "00000000-0000-0000-0000-000000000000";

export function serializeRelayFrame(f: RelayFrame): Uint8Array {
  const srcU = utf8(f.sourceUhid ?? "");
  const dstU = utf8(f.destinationUhid ?? "");
  const relayU = utf8(f.relayUhid ?? "");
  const payload = f.payload ?? new Uint8Array(0);

  const size =
    1 + 1 + 1 +
    2 + srcU.length + 2 + dstU.length + 2 + relayU.length +
    16 + 8 + 4 + 8 +
    4 + payload.length;

  const buf = new Uint8Array(size);
  const dv = new DataView(buf.buffer);
  let o = 0;
  buf[o++] = RELAY_FRAME_VERSION;
  buf[o++] = f.type & 0xff;
  buf[o++] = (f.status ?? RelayStatus.Ok) & 0xff;
  o = writeStr(buf, dv, o, srcU);
  o = writeStr(buf, dv, o, dstU);
  o = writeStr(buf, dv, o, relayU);
  buf.set(uuidToBytes(f.connectionId && f.connectionId.length > 0 ? f.connectionId : NIL_UUID), o); o += 16;
  dv.setBigInt64(o, BigInt(f.reservationExpiresAtMs ?? 0), true); o += 8;
  dv.setInt32(o, f.limitDurationSeconds ?? 0, true); o += 4;
  dv.setBigInt64(o, BigInt(f.limitDataBytes ?? 0), true); o += 8;
  if (payload.length > MAX_RELAY_PAYLOAD) throw new Error("Relay payload too large");
  dv.setInt32(o, payload.length, true); o += 4;
  buf.set(payload, o);
  return buf;
}

export function deserializeRelayFrame(data: Uint8Array): RelayFrame {
  const r = new Reader(data);
  r.expectVersion();
  const type = r.u8();
  if (type === 0 || type > RelayMessageType.Data) throw new Error(`Relay: invalid message type ${type}`);
  const status = r.u8();
  if (status > RelayStatus.MalformedMessage) throw new Error(`Relay: invalid status ${status}`);
  const sourceUhid = r.str();
  const destinationUhid = r.str();
  const relayUhid = r.str();
  const connectionId = r.uuid();
  const reservationExpiresAtMs = r.i64();
  const limitDurationSeconds = r.i32();
  const limitDataBytes = r.i64();
  const payload = r.bytes32();
  return {
    type: type as RelayMessageType,
    status: status as RelayStatus,
    sourceUhid,
    destinationUhid,
    relayUhid,
    connectionId,
    reservationExpiresAtMs,
    limitDurationSeconds,
    limitDataBytes,
    payload,
  };
}

// ── Low-level helpers (identical idiom to DtnEnvelope) ───────────────────────

function utf8(s: string): Uint8Array {
  return new TextEncoder().encode(s);
}

function writeStr(buf: Uint8Array, dv: DataView, o: number, bytes: Uint8Array): number {
  if (bytes.length > 65535) throw new Error("Relay string too long");
  dv.setUint16(o, bytes.length, true); o += 2;
  buf.set(bytes, o); o += bytes.length;
  return o;
}

function uuidToBytes(uuidStr: string): Uint8Array {
  const hex = uuidStr.replace(/-/g, "");
  const bytes = new Uint8Array(16);
  for (let i = 0; i < 16; i++) bytes[i] = parseInt(hex.substr(i * 2, 2), 16);
  return bytes;
}

function bytesToUuid(bytes: Uint8Array): string {
  const hex = Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}

class Reader {
  private readonly dv: DataView;
  private o = 0;
  constructor(private readonly data: Uint8Array) {
    this.dv = new DataView(data.buffer, data.byteOffset, data.length);
  }
  expectVersion(): void {
    const v = this.u8();
    if (v !== RELAY_FRAME_VERSION) throw new Error(`Relay: unsupported frame version 0x${v.toString(16)}`);
  }
  u8(): number {
    return this.data[this.o++];
  }
  uuid(): string {
    const s = this.data.slice(this.o, this.o + 16);
    this.o += 16;
    return bytesToUuid(s);
  }
  i32(): number {
    const v = this.dv.getInt32(this.o, true);
    this.o += 4;
    return v;
  }
  i64(): number {
    const v = this.dv.getBigInt64(this.o, true);
    this.o += 8;
    return Number(v);
  }
  u16(): number {
    const v = this.dv.getUint16(this.o, true);
    this.o += 2;
    return v;
  }
  str(): string {
    const n = this.u16();
    const s = new TextDecoder().decode(this.data.slice(this.o, this.o + n));
    this.o += n;
    return s;
  }
  bytes32(): Uint8Array {
    const n = this.i32();
    if (n < 0 || n > MAX_RELAY_PAYLOAD) throw new Error(`Relay: invalid payload length ${n}`);
    const b = this.data.slice(this.o, this.o + n);
    this.o += n;
    return b;
  }
}
